Chap_API_Event.tex

```diff
@@ -41,13 +41,17 @@ \section{Notification and Management}
+Users can specify the callback order of a handler within its category at the time of registration. Ordering can be specified either by providing the relevant returned event handler registration ID or using event handler names, if the user specified an event handler name when registering the corresponding event. Thus, users can specify that a given handler be executed before or after another handler should both handlers appear in an event chain (the ordering is ignored if the other handler isn't included). Note that ordering does not imply immediate relationships. For example, multiple handlers registered to be serviced after event handler \textit{A} will all be executed after \textit{A}, but are not guaranteed to be executed in any particular order amongst themselves.
 
 In addition, one event handler can be declared as the \textit{first} handler to be executed in the chain. This handler will \textit{always} be called prior to any other handler, regardless of category, provided the incoming event matches both the specified range and event code. Only one handler can be so designated --- attempts to designate additional handlers as \textit{first} will return an error. Deregistration of the declared \textit{first} handler will re-open the position for subsequent assignment.
 
 Similarly, one event handler can be declared as the \textit{last} handler to be executed in the chain. Note that this handler will not be called if the chain is terminated by an earlier handler. Only one handler can be designated as \textit{last} --- attempts to designate additional handlers as \textit{last} will return an error. Deregistration of the declared \textit{last} handler will re-open the position for subsequent assignment.
 
-Upon completing its work and prior to returning, each handler \textit{must} call the event handler completion function provided when it was invoked (including a status code plus any information to be passed to later handlers) so that the chain can continue being progressed. \ac{PMIx} automatically aggregates the status and any results of each handler (as provided in the completion callback) to that from all prior handlers so that each step in the chain has full knowledge of what preceded it. An event handler can terminate all further progress along the chain by passing the \refconst{PMIX_EVENT_ACTION_COMPLETE} status to the completion callback function.
+\adviceuserstart
+Note that \textit{first} and \textit{last} handlers are qualified within their category. That is, default event handlers will always be invoked after a single code event handler, even when the later is qualified as \textit{last}.
+\adviceuserend
```

Comment:
This is not actually correct - the first and last handlers are absolute and not within a category, and the last handler will be invoked after the default so it is truly the _last_ one.

Reply:
Maybe the correct question here is: we don't do what you describe today - should we? It would be a significant change to the current code, so it would be nice to first understand what is gained by doing it. Also need to clarify things like: is the "first" handler for the single code case different for each code? Or is it the "first" handler for all single-code registrations?

Could get awfully complex...

Reply:
My bad. I misread the paragraph immediately above. Still a good thing that we review this piece.

I have no particular problem with the 'first' attribute being 'first' for all categories. 

I have a concern with the 'last' attribute. Given that it executes after the 'default' handler, and that the default handler often will abort the job, that could be of limited usefulness, or am I missing a point?

Reply:
My expectation was that the "abort" handler would be the one specified to execute last, but maybe that doesn't match what others would think. I can see where someone might feel that a "default" handler should play that role. Perhaps a "last" handler doesn't make sense?